server/src/server.ts

```diff
@@ -179,7 +179,10 @@ export default class Server {
       this.print(`${getElapsedSeconds(start)}s to perform maintenance`);
     }
 
-    // await migrate(this.model.db.mysql.conn);
+    if (this.config.ENABLE_MIGRATIONS) {
+      await migrate(this.model.db.mysql.conn);
+      this.model.isMigrated = true;
```

Comment:
why are we setting `isMigrated` outside of model? shouldn't model keep this internally, and set this internally?

Reply:
Would you prefer `server` calling `model.setIsMigrated(true)`. I'm agnostic about it tbh.

Reply:
I know it's too late now, but I'll add some general thoughts.

Right now we are mixing OOP approaches with more functional approach, which is why the above is confusing. We have this function that we import (without being attached to any object) which modifies state, and now we have to manually keep track of the fact that we modified that somewhere. Setting the `isMigrated` flag on the model object explicitly is a symptom of this problem, and it not good OOP practice (more about that in a bit).

Better would be to do something like this:
```
this.mode.migrate();
```
which internally to model will set a private `isMigrated` flag after it does it's work. So then elsewhere in server you could call something like `this.model.isMigrated()` to check if the migration happened. 

I know this involves forwarding calls (ie. `migrate` in this case) through multiple objects, which you may not like. But it is much more readable that mixing functional and OOP approaches, and also setting flags on an object when an object has no idea about that property. As we sit now, we are breaking OOP division of responsibilities by having isMigrated on model, because model has no idea what that is or where it comes from. This will make it harder for other developers to read and figure out what we have done and why we have done it.

I'm not pushing back too hard yet, because we plan to rip out a bunch of this stuff once the DB is in a happy place. But I think this is something to keep in mind going forward.

Reply:
It's never too late and I'm always happy to talk about clarity of code 👍 

I only partially agree that model having a isMigrated flag violates division of responsibilities as model in its current form has a larger set of (high-level) responsibilities under which `isMigrated` could reasonably fall. Also I don't quite see how changing a setter to a method call would unviolate that principle. And we still need to inform `clips.ts` of the state of the migration as that is where the branching happens.

Would the OOP solution be to have an object with a migration method and a flag-property that indicates whether the migration has completed (and then pass that wherever branching needs to occur)? I can see doing it that way, when the migration would result in multiple properties being set.

Reply:
The flag itself is not violating division of responsibilities, and I agree private member var with accessor is same as setter.

What's violating division of responsibilities in this case is that `server` invokes a migrate function it imported from db within the model, and then explicitly sets isMigrate. To adhere to division of responsibilities, migrate should happen within the model, and the model should set it's own `isMigrated` flag. Having `server` play such a conductor role where migrate happens without going directly through the model, and modle is only used as a namespace to store a flag about migration is not a pattern i've seen before, and I'd venture to say it's not good practice.

Speaking of `clips`, since it lives inside the model, the model should inform clips if and when migration happens. As of now, it's unclear how much control/responsibilities the model has over migration, and in my opinion it should be the conductor, not `server`.

Generally OOP is a nice way to say who does what. I think the current setup it's hard, because sometime we take an object/state approach, and sometimes we are taking a more functional approach, and we should aim for consistency. Since OOP is what we have been using from the ground up, I feel like we should adhere to that. In a very specific sense, this means that model should probably do the migration, and set it's own `isMigrated` flag, and also inform sub-components when and if migration has occurred.

Reply:
It may seem like I'm speaking a lot about something that is essentially a nitpick. And I can understand that point of view. But in reality, this little thing has important ramifications.

I think the way we are doing migration, ie. having `server` import and invoke a function from deep within the model subcomponents, and then after completion setting a flag on a different subcomponent breaks OOP convention (for reasons explained above). In the short term, it's easy to see what's going on and so it no big deal. But as time goes on if we keep allowing anti-conventions to slip in, the project becomes much harder to reason about. And will be unfriendly to new people on the project.

Reply:
I think I probably didn't reason too much about who's responsibility it is, because when I think about it now I'd agree with you that it feels more natural to put it in the model (which would probably forward the call to the db). The reason I've done it was probably a pragmatic one: The API needs to know about the successful migration as well. So if we move it into the model (which we probably should do for separation of concerns reasons), we'd need to pass a migration-complete callback into the model. But while that's something worth mentioning and I wouldn't shy away from doing the thing that makes more sense even if it complicates things a bit.

> It may seem like I'm speaking a lot about something that is essentially a nitpick. And I can understand that point of view. But in reality, this little thing has important ramifications.

Not at all. I think it's great we can have these kinds of discussions. Imo as soon as something seems off to someone, it's legitimate to talk about it. The best case for me is that I learn something, which isn't that bad of a thing for me. 😁